File: python_cpp/uPIMulator_backend/src/initializer/int_initializer.cc
#include "initializer/int_initializer.h"

#include <cassert>
#include <cmath>
#include <random>

namespace upmem_sim::initializer {

int64_t IntInitializer::value_by_range(int64_t min_value, int64_t max_value) {
  assert(min_value < max_value);

  std::random_device random_device;
  std::mt19937 generator(random_device());
  std::uniform_int_distribution<int64_t> distribution;

  int64_t range = max_value - min_value;
  return distribution(generator) % range + min_value;
}

int64_t IntInitializer::value_by_width(abi::word::Representation representation,
                                       int width) {
  if (representation == abi::word::UNSIGNED) {
    return value_by_range(0, static_cast<int64_t>(pow(2, width)));
  } else {
    return value_by_range(-static_cast<int64_t>(pow(2, width - 1)),
                          static_cast<int64_t>(pow(2, width - 1)));
  }
}

}  // namespace upmem_sim::initializer

File: python_cpp/uPIMulator_frontend/src/initializer/int_initializer.py
import random
from typing import Any, List

from abi.word.representation import Representation


class IntInitializer:
    def __init__(self):
        pass

    @staticmethod
    def value_by_range(min_value: int, max_value: int) -> int:
        assert min_value < max_value

        return random.randrange(min_value, max_value)

    @staticmethod
    def value_by_width(representation: Representation, width: int):
        assert width > 0

        if representation == Representation.UNSIGNED:
            return IntInitializer.value_by_range(0, 2 ** width)
        elif representation == Representation.SIGNED:
            return IntInitializer.value_by_range(-(2 ** (width - 1)), 2 ** (width - 1))
        else:
            raise ValueError

    @staticmethod
    def value_by_list(list_: List[Any]) -> Any:
        return random.choice(list_)
